### ISSUE TITLE
Consider to add an abs package manager

### ISSUE THREAD

**Issue by eagleusb** [abs-lang/abs#214]
Hello,

first and foremost the ABS seems fun and practical to code with, thanks for that.

By experience, language with a good associated package manager are easier to adopt for obvious reasons, that's why I would love to code, package and reuse some libraries made with ABS.

It's also something really missing with Bash.

What do you think ?

My 2 cents,

**Comment by odino** [comment]
Definitely on the wishlist. The main obstacle right now is defining the structure of the package manager, the commands it can run etc. I am inclined to implement something extremely simple:

* `abs get github.com/user/pkg`
* Installs in a local `vendor` folder by cloning the repo
* then you can require `vendor/user/pkg`

Later on one can add support for lockfiles, sources other than girt repos etc etc but as a first step I think this could be something.

@ntwrick your input is appreciated as usual :)

**Comment by gooodmorningopenstack** [comment]
Awesome.
While on the news (Microsoft github org.), it reminds me that the path of the package should be normalized (lowercased...), in order to avoid a possible mess if renamed.

My 2 cents,

**Comment by ntwrick** [comment]
You can also write a short ABS snippet that tests for the existence of your library package directory (git or otherwise) and then either `clone a repo` or `download and untar an archive` as needed. This can be coded in a few lines as part of your `~/.absrc file` where you would then `source(file)` the file(s) containing the library functions and environment variables into the running ABS environment. 

This will give you total control over how you package and distribute all your ABS, BASH, or other shell library functions and does not require you to adopt an ABS-specific distribution model.

**Comment by odino** [comment]
The downside with the snippet is that it adds a bit of a barrier for people who are not familiar with ABS itself. I think `abs get github.com/user/package` could work well and be minimal enough.

**Comment by odino** [comment]
I would like to get this done before introducing try / catch, so that we can start writing part of the std library as an "external" module, eg:

```
errors = require("errors")

return errors.new(xyz)
```

I think the basic feature that a package manager would have to have:

* `abs install github.com/user/repo`
* clone the repo in a `./vendor` folder (and I'm ok if we just support Git / GH for now)
* update the `require` function to be able to distinguish between 3rd party packages and built-in modules (standard library)

Once this is done, we can start thinking about try / catch, built-in structure like mutexes / semaphores etc.

**Comment by mingwho** [comment]
We need to think about a way to distinguish external modules when we require them.
E.g.
```
errors = require("errors")
return errors.new(xyz)
```
Here "errors" could be a local file in the same directory, or an external module in the `./vendor` folder, or a built-in module. Maybe we should enforce the use of `require(./vendor/user/pkg)` for external modules?

**Comment by odino** [comment]
Multiple ways to tackle this but I haven't thought of the best one in my head (all following examples are vendor vs local package):

* `package` vs `./package`
* `@package` vs `package`
* `package` vs `github.com/user/package`
* `:package` vs `package`

My slight bias would be towards option 1, as I'm used to it. Option 3 (like golang does) is also not a bad way to think about it but might be overkill, and it's definitely less pragmatic / more verbose. Since we want to make sure ABS doesn't really force users to type that much, I'd lean on no.1 for now. Something to think about is naming collision, eg. if I `abs install github.com/user/util` and then `abs install github.com/another_user/util`, there should be a way to avoid this -- here my suggestion would be to allow users to specify the full path when package names collide. And in the future we can think of a `.lock` file that we use to sort these collisions out, where the file looks like:

``` json
[
  {"package": "github.com/user/util", "version": "xyz", "alias": "util"}
]
```

and when you try to `abs install github.com/another_user/util` the CLI tells you "there is another package aliased to `util`, please pick a new alias for this package or press enter to overwrite the dependency". Fun stuff we can think about later on... :smile: 

Also one of the things I think we need to fix is scoping in modules. If a module sets a variable we shouldn't have that variable set at global level. This does not necessarily have to be fixed in this PR but, if you want, it's a nice to have. Things we would need to change are:

* differentiate how we do [source vs require](https://github.com/abs-lang/abs/blob/master/evaluator/functions.go#L324-L332): source can pollute the global environment (like in a shell) but require shouldn't
* fix the [docs](https://www.abs-lang.org/types/builtin-function)
* replace the [global environment for require calls](https://github.com/abs-lang/abs/blob/master/evaluator/functions.go#L1614): we instead need to use an [enclosed environment](https://github.com/abs-lang/abs/blob/master/object/environment.go#L12-L16), which is what we do when a function `f() { ... }` [is called](https://github.com/abs-lang/abs/blob/cc2191ee9911c045dffda8a58cbb2b9cdef59b2e/evaluator/evaluator.go#L1046) (the function inherits the environment but cannot modify the global one eg. set new variables)

Let me know if you'd like me to have a look at this instead. It might get too beefy to work on the package manager as well as differentiating source/require at the same time.
